### ISSUE TITLE
Virtual library for mirage-crypto-pk

### ISSUE THREAD

**Issue by jonahbeckford** [mirage/mirage-crypto#197]
TLDR: Having a virtual library so that GMP is not a hard-dependency of mirage-crypto-pk.

This was forked off from https://github.com/mirage/mirage-crypto/pull/137.

The specific request:

- Create a virtual library (https://dune.readthedocs.io/en/stable/variants.html). Let's call it `mirage-crypto-bigint` for now.
- Inside it has a `mirage_crypto_bigint.mli` specifying exactly what types and functions `mirage-crypto` needs from Zarith.
- Make an implementation library `mirage-crypto-bigint.zarith` that just passes through to Zarith.

With a virtual library we could use the slower but much more portable `num` to implement the virtual library (ie. `mirage-crypto-bigint.num`). That would allow a slow Windows implementation. That slowness would be fine for many use cases (ex. testing out a web server on a Windows desktop) since I think the common use case is to deploy TLS endpoints on Linux. *I believe TLS is the main use case for mirage-crypto-pk.*

Someone would also be able to implement a new constant-time bignum implementation. I might even port a bignum from Rust or bind another C bignum library myself eventually, but that is a small chance since for most of my needs I prefer TLS offloading. (Yet if I had to use in-process TLS I would avoid the GMP library both because of its license and how difficult GNU software is to build with the MSVC compiler)

**Comment by kit-ty-kate** [comment]
I too would be happy to see the dependency on GMP go.
@hannesm has there been any benchmarks comparing performance of mirage-crypto when using `zarith` vs. `num`? (or even vs. some pure OCaml implementation optionally for good measure)

**Comment by hannesm** [comment]
> I too would be happy to see the dependency on GMP go. [@hannesm](https://github.com/hannesm) has there been any benchmarks comparing performance of mirage-crypto when using `zarith` vs. `num`? (or even vs. some pure OCaml implementation optionally for good measure)

Not that I'm aware of. But as far as I remember, Num doesn't provide the functionality needed (but I may be wrong).
